compiler/parser/parser.cpp

```diff
@@ -505,11 +508,41 @@ static ModuleSymbol* parseMod(const char* modName, bool isInternal) {
 static bool containsOnlyModules(BlockStmt* block, const char* path);
 static void addModuleToDoneList(ModuleSymbol* module);
 
+//
+// This is a poor person's check to see whether we've already parsed
+// this file before to avoid re-parsing the same thing twice.  It'd be
+// nice to upgrade to the C++17 filesystem library to do this, but we
+// haven't done portability testing of those features and it's close
+// to the release as I add this.  Instead, check to see whether the
+// path to the file matches one we've parsed before, adding special
+// handling to paths with the form './abc.chpl'.  This will catch
+// common cases for how we build up filenames using search paths,
+// though other cases will definitely slip past.
```

Comment:
Our C runtime has support for this:

https://github.com/chapel-lang/chapel/blob/da7ecc219a5598b250832caa6c41961c5bf33c24/runtime/src/chpl-file-utils.c#L335-L351

It seems to me that we could call the same `realpath` function from the compiler without needing any particular C++ thing.

Reply:
Good idea, thanks for the pointer.  Looks like we already have a wrapper for it in the compiler as well.